src/boot/efi/efi-string.c

```diff
@@ -272,6 +274,419 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+
+        if (is_signed && (int64_t) u < 0) {
+                buf[0] = '-';
+                buf++;
+
+                int rem = -((int64_t) u % base);
+                u = (int64_t) u / -base;
+                tmp[n++] = digits[rem];
+        }
+
+        while (u > 0 || n == 0) {
+                int rem = u % base;
+                u /= base;
+                tmp[n++] = digits[rem];
+        }
+
+        assert(n + min_digits + 5 < SCRATCH_BUF_SIZE);
+
+        if (hex_prefix) {
+                buf[0] = '0';
+                buf[1] = lower ? 'x' : 'X';
+                buf += 2;
+        }
+
+        while (min_digits > n) {
+                buf[0] = '0';
+                buf++;
+                min_digits--;
+        }
+
+        while (n > 0) {
+                *buf = tmp[--n];
+                buf++;
+        }
+
+        *buf = '\0';
+}
+
+static const char *parse_format_specifier(
+                EFI_STATUS status,
+                const char *format,
+                va_list *ap,
+                char16_t tmp[SCRATCH_BUF_SIZE],
+                bool *ret_pad_zero,
+                int *ret_precision,
+                int *ret_field_width,
+                const char **ret_str8,
+                const char16_t **ret_str16,
+                const wchar_t **ret_wstr) {
+
+        assert(ap);
+        assert(ret_pad_zero);
+        assert(ret_precision);
+        assert(ret_field_width);
+        assert(ret_str8);
+        assert(ret_str16);
+        assert(ret_wstr);
+
+        bool alternative_form = false, have_field_width = false, long_arg = false;
+        size_t field_length = sizeof(int);
+
+        for (;; format++)
+                switch (*format) {
+                case '0':
+                        *ret_pad_zero = true;
+                        break;
+                case '.':
+                        have_field_width = true;
+                        break;
+                case '#':
+                        alternative_form = true;
+                        break;
+
+                case '*':
+                case '1' ... '9': {
+                        uint64_t u;
+
+                        if (*format == '*')
+                                u = va_arg(*ap, int);
+                        else {
+                                if (!parse_number8(format, &u, &format) || u > INT_MAX)
+                                        assert_not_reached();
+                                format--; /* Point it back to the last digit. */
+                        }
+
+                        if (have_field_width && *ret_precision < 0)
+                                *ret_precision = u;
+                        else if (!have_field_width && *ret_field_width < 0)
+                                *ret_field_width = u;
+                        else
+                                assert_not_reached();
+                        break;
+                }
+
+                case '%':
+                        *ret_str16 = u"%";
```

Comment:
What happens when field width or precision is specified? E.g. `%9.9%`??

Reply:
gcc will complain about any of these: `%9.9% %9.% %0% %.0% %#%`